pixels-turbo/pixels-worker-vhive/src/main/java/io/pixelsdb/pixels/worker/vhive/PixelsWriterStreamImpl.java

```diff
@@ -0,0 +1,830 @@
+             */
+            // writer.reset();  // This seems to be slower than creating a new writer for each row group.
+            columnWriters[i] = newColumnWriter(children.get(i), columnWriterOption);
+        }
+
+        if (rowGroupDataLength + 32 < recordedRowGroupDataLen || rowGroupDataLength + 8 > recordedRowGroupDataLen)  // XXX: if (rowGroupDataLength != recordedRowGroupDataLen)
+            throw new IOException("The calculated rowGroupDataLength is not equal to the recorded value");
+
+        if (partitioned) {
+            // partitionColumnIds has been checked to be present in the builder.
+            curPartitionInfo.addAllColumnIds(partKeyColumnIds.orElse(null));
+            curPartitionInfo.setHashValue(currHashValue);
+        }
+
+        // put curRowGroupIndex into rowGroupFooter
+        StreamProto.StreamRowGroupFooter.Builder rowGroupFooterBuilder =
+                StreamProto.StreamRowGroupFooter.newBuilder()
+                        .setRowGroupIndexEntry(curRowGroupIndex.build())
+                        .setRowGroupEncoding(curRowGroupEncoding.build())
+                        .setNumberOfRows(curRowGroupNumOfRows);
+        if (partitioned) {
+            rowGroupFooterBuilder.setPartitionInfo(curPartitionInfo.build());
+        }
+        StreamProto.StreamRowGroupFooter rowGroupFooter = rowGroupFooterBuilder.build();
+        // XXX: rowGroupIndex and rowGroupEncoding are the same for all row groups in the same file?
+        //  If so, we can send them only once per file.  -- Row group footer accounts for <1% of the total data size. No need to optimize.
+
+        // write and flush row group footer
+        byte[] footerBuffer = rowGroupFooter.toByteArray();
+        curRowGroupFooterOffset = byteBuf.writerIndex();
+        byteBuf.writeInt(footerBuffer.length);
+        byteBuf.writeBytes(footerBuffer, 0, footerBuffer.length);
+        writtenBytes += footerBuffer.length;
+
+        this.fileRowNum += curRowGroupNumOfRows;
+        this.rowGroupNum++;
+
+        // Send row group to server (an additional step compared to PixelsWriterImpl)
+        if (!partitioned && uri == null) {
+            uri = URI.create(fileNameToUri(fileName));
+        }
+        logger.debug("Sending row group with length: " + byteBuf.writerIndex() + " to endpoint: " + (partitioned ? fileNameToUri(fileNames.get(currHashValue)) : uri.toString()));
+        Request req = httpClient.preparePost(partitioned ? fileNameToUri(fileNames.get(currHashValue)) : uri.toString())
+                .setBody(byteBuf.nioBuffer())
+                .addHeader("X-Partition-Id", String.valueOf(partitionId))
+                .addHeader(CONTENT_TYPE, "application/x-protobuf")
+                .addHeader(CONTENT_LENGTH, byteBuf.readableBytes())
+                .addHeader(CONNECTION, "keep-alive")
+                .build();
+
+        // DESIGN: We do not implement async retry here, because experiment shows that the bottleneck is in the sequential
```

Comment:
Why do we need a retry here?

Reply:
Because the timings between threads are complicated. Previous experiments show that sometimes the client would throw an exception and crash because it attempts to connect to a server that hasn't started yet.
If it is discouraged to add a retry here, we can try to add more timing control, e.g. to make `StreamWorkerCommon.getPort()` blocking until the server is up.